WIP 42 Initial description of Module-based hexagonal component architecture (#45)
devonfw/java: modules/ROOT/pages/architecture/hexagonale_architecture.adoc

Therefore, we propose to implement the domain model as anemic entities and make usage of use cases and services to implement the business logic and interact with the domain models.

== Module-based hexagonal component architecture
The previously described approach of structuring using package structures is particularly suitable for lightweight and smaller applications.

For larger applications, it makes sense to additionally implement the division using Maven or Gradle modules.

> This has the advantage that the application core of the hexagon remains completely without special technical frameworks such as JPA.

Reviewer: From my point of view, I would structure the article a bit differently. For me this sentences of the technoogy framework free core comes a bit out of nowhere. 
I would suggest the following:

1. Start with the business component division. Why have we choosen it (mainly for a future proof structure when more components arise during time)
2. Then the division into core and adapter. Here answering why diving them again (For me here the technology free core belongs) 
3. The bootstrap module at the end. Explaining that it's necessary to define it here, so that the starting point of the applicaiton has the change to collect all dependencies necessary for the DI framework are collected. Hence it needs dependencies to all modules available.

The division into modules looks like this:

[source,plaintext]
----
application A1
├── bootstrap (module)
└── component C1 (module)
    ├── core (module)
    │   ├── domain
    │   ├── port
    │   │   ├── in
    │   │   └── out
    │   ├── usecase
    │   └── service (optional)
    └── adapter (module)
        ├── in
        │   └── web
        │       └── (analog with packages)
        └── out
            └── repository
                └── (analog with packages)
----

At the main level of the application there is the Bootstrap Module, which in Spring contains, for example, the configurations and SpringBootApplication.
This module provides the starting point of the application and at the same time has a dependency on all cores and adapters of the business components.
The business components, of which there can be any number, are also defined as modules at the main level.
In the first step, we only assume one business module.
This is divided into both the core module and the adapter module.
In this way, core and adapter can define different dependencies and the core remains free of technical frameworks.

> The adapters in turn are divided into `in` and `out` adapters as in the package structure described above.

Reviewer: This sentence ist kind of irritating as we're talking about dividing it further and then it's saif only on packages. I would turn it around and say that we recommend to group all adapters in one module, because reasons.
Then argue that this adapter module uses packages to structure the different adapters.

However, no division into modules is recommended here, but instead exclusively into packages.
A division here into modules increases the complexity significantly and also makes the expansion of new adapters much more complex, as these also have to be made known to the bootstrap module.
With a large number of adapters and a high number of different dependencies, a tradeoff is to further divide the in and out adapters into modules.
However, the recommendation here is to initially use packages.

> The management of dependencies should aim to define them as far "down" as possible.

Reviewer: For me it's clear what's meant here. Maybe an example would be good. like the JPA dependency should be at the adapter module. THe version number can be in the root module. Maybe even reason why this is necessary. (If the JPA dependency is in the root module even teh core can access and use it)

This means in particular that technical dependencies are always defined in the adapters.
In order to keep versions consistent if possible, dependency management in the master pom can be used in Maven.
The advantage of this is that the version only needs to be defined at the main level and does not need to be specified in every adapter pom.

> For internal dependencies it is necessary that the adapter has access to the core.

Reviewer: Internal dependencies sounds like dependencies betweeen components. Maybe it would be better to explain this a bit further. Like "It is necessary that the adapters can access the core, so that incoming adapters can call core logic and the dependecy inversion mentioned earlier between outbound adapters and core can be implemented. hence there need to be a dependency between adapter module and the core module."

Therefore, we propose to implement the domain model as anemic entities and make usage of use cases and services to implement the business logic and interact with the domain models.

== Module-based hexagonal component architecture

> The previously described approach of structuring using package structures is particularly suitable for lightweight and smaller applications.

Reviewer: ```suggestion
The previously described approach of using packages to maintain a hexagonal architecture is particularly suitable for lightweight and smaller applications.
```

For larger applications, it makes sense to additionally implement the division using Maven or Gradle modules.
This has the advantage that the application core of the hexagon remains completely without special technical frameworks such as JPA.
As a result, it is technically not possible to use these functions in the core, which permanently leads to a "cleaner" core.

//[[img-t-hexagonal-architecture-module]]
//.Module-based hexagonal component architecture

> //image::hexagonal_component_architecture_modules.drawio.svg["devonfw hexagonal architecture module",scaledwidth="80%",align="center"]

Reviewer: Why is this a comment? 
If you think the image is not bringing a real benefit I would rather remove it here and in the images directory to keep it clean.

The division into modules looks like this:

[source,plaintext]
----
application A1
├── bootstrap (module)

> └── component C1 (module)

Reviewer: maybe call ist business component to stay aligned with the text.

        └── out
            └── repository
                └── (analog with packages)
----

Within the application we structure the individual business components using modules.
The main purpose of this is to have a structure that is future-proof and allows multiple business components to be integrated.
Modules at this level have the advantage that they can be developed independently and communication must be explicitly regulated.
For the sake of simplicity, we initially only assume one business component.
The business component is in turn divided into two further modules `core` and `adapter`.
This separation allows the dependencies for both modules to be independent of each other and the core to remain free of technical frameworks.
The direction of communication is also clearly regulated by the dependencies between core and adapter.
In addition to the module for the business component, there is a `bootstrap` module that can be seen as a starting point for the application.
When using Spring Boot, the `SpringBootApplication` and configurations can also be found here.
This module in turn has a dependency on all business components (to be precise on their core and adapter).

> With regard to the adapters, the recommendation is to create them within one module in the package structure mentioned above.
> This offers a good compromise between structuring and tailor-made dependencies and manageable complexity on the other hand.

Reviewer: I had to read the two sentences a few times to understand the meaning. 
Maybe splitting it up into multiple sentences makes it a bit clearer.
```suggestion
The adapter module contains all incoming and outgoing adapters of a business component. 
This offers a good compromise between structuring and tailor-made dependencies and manageable complexity on the other hand.
Therefore, we suggest to structure the adapter module using packages to differentiate incoming and outgoing adapters and on the next level the specifics of adapters like web, repository, etc.
See the package project structure to get the full picture. 
```

Given the usual level of complexity, it can be assumed that there are usually no more than 2-3 adapters, so structuring using packages is sufficient here.
As complexity increases, two additional options become available:

1. Separate modules for in and out. This allows the frameworks for incoming and outgoing adapters to be maintained separately and provides an additional level of structuring.
2. Individual modules for each adapter. This means that each adapter can be viewed and developed completely isolated.
The dependencies can also be maintained independently.

> With both options, the complexity increases accordingly.

Reviewer: The complexity of what? Because complexity is also increasing 6 lines above. 
I would mention the maintenance of modules and explicit depenednecies for example.

The recommendation is to start with a package structure and divide it further if necessary.

> The management of dependencies should aim to define them as far "down" as possible and not on the root or business component level.

Reviewer: ```suggestion
Dependencies should be defined as far "down" as possible and not on the root or business component level.
```

In particular, technical frameworks such as JPA should be defined directly in the adapter module so that they are not available in the core module.

> In order to keep versions consistent, dependency management in the master pom can be used in Maven.

Reviewer: Hier wäre ein Link zu Dependency Management in Maven hilfreich für alle, die nicht direkt wissen was gemeint ist.

The advantage of this is that the version only needs to be defined at the main level and does not need to be specified in every adapter pom.
This approach also prevents the same framework from being used in different versions in the adapters.

It is necessary that the adapters can access the core and therefore have a dependency so that, on the one hand, incoming adapters can call logic in the core and, on the other hand, outgoing adapters from the core can be used.
The latter is only possible via dependency inversion, as the core is not allowed to access the outgoing adapter directly.
In this case, the out ports in the core are defined and used as an interface.
The dependency injection framework in Spring or Quarkus then determines in the background the appropriate implementation from the out adapters that implement the out port interface.
In this way, it is possible for the core to call functionality in the adapters without having a dependency in this direction.
Under no circumstances should there be a dependency from the core to the adapter.

> This would mean that fundamental aspects of the hexagonal architecture could be undermined and technical frameworks could also be used in the core.

Reviewer: I would remove this sentence. If there's a dependency then it's no hex arch anymore. This one direction change is the fundamental part. Technical frameworks can be used as well and with spring etc probably are used in the core (at least annotations like validation and JEE stuff). Therefore, this sentence can either be reduced to "This would undermine the fundamental principle of a hexagonal architecture and destroy the benefits of it" or it could be left out and we assume that the reader will understand the implication

NOTE: In case you are using Quarkus environment the dependency inversion is not possible out of the box.
That means that outgoing ports cannot injected as expected which causes errors during application start.

> In that case you can inject outgoing ports with `Instance<YourOutgoingPort>` which decouples the injection process.

Reviewer: Maybe a link to the issue or documentation describing this would be helpful in case it changes in the future.

Author: its not really an issue, its expected behavior: https://quarkus.io/guides/cdi#hm-wait-a-minute-what-happens-if-multiple-beans-declare-the-same-type

Will link it in the text.

        │           └── BookingComponentC1InAdapter.java
        └── out
            └── repository
                └── ...
----

In this way, the two business components can communicate with each other within a modular architecture.
At first glance, this seems complex and may involve overhead, since, for example, two mappings have to be implemented.
However, this is deliberately chosen because, on the one hand, it means that communication between business components is used consciously and only where really necessary.
On the other hand, these adapters represent a predetermined breaking point for later division.
More on this in the next section.

=== Extract business components into microservices

Increasing technical complexity or changing scaling requirements may make it necessary to (partially) convert the existing module structure into a microservice architecture over time.
The previously selected structure of the hexagonal architecture and the explicit communication via JVM adapters enable such a transformation with relatively little effort.
The decisive advantage is that communication takes place beforehand via firmly defined interfaces.

> These now have to be switched from a communication technology based purely on JVM means to a communication technology.

Reviewer: I think it makes sense to explain this pure JVM techniques with an example.
E.g. direct procedure call from the outgoung adapter C1 to the incoming adpter of C2
or a internal messaging provider like... Quarkus or Sprin possibilities here
